core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java

```diff
@@ -87,6 +108,34 @@ public static BaseFileSystem create(FileSystemContext context) {
    */
   protected BaseFileSystem(FileSystemContext fsContext) {
     mFsContext = fsContext;
+    mBlockStore = AlluxioBlockStore.create(fsContext);
+  }
+
+  /**
+   * Shuts down the FileSystem. Closes all thread pools and resources used to perform operations. If
+   * any operations are called after closing the context the behavior is undefined.
+   *
+   * @throws IOException
+   */
+  @Override
+  public void close() throws IOException {
+    if (!mClosed) {
+      synchronized (mCloseLock) {
+        // TODO(zac) Determine the behavior when closing the context during operations.
+        if (!mClosed) {
+          mClosed = true;
+          Factory.FILESYSTEM_CACHE.remove(new FileSystemKey(mFsContext.getClientContext()));
+          mFsContext.close();
+        }
+      }
+    }
+  }
+
+  @Override
+  public boolean isClosed() {
+    synchronized (mCloseLock) {
+      return mClosed;
+    }
```

Comment:
What does synchronizing here accomplish?

Reply:
Originally I had put it there because I didn't want this to return if the cache was in the process of closing. However given that `close()` marks `mClosed` as true before doing anything the synchronization doesn't add any value here. I'll remove it

Reply:
The synchronized is necessary if `mClosed` is non-volatile.

Reply:
I see, thanks for the explanation.